sdk/src/trace/tracer_provider.cc

```diff
@@ -19,11 +19,28 @@ TracerProvider::TracerProvider(std::unique_ptr<SpanProcessor> processor,
                                                      std::move(id_generator)))
 {}
 
-opentelemetry::nostd::shared_ptr<opentelemetry::trace::Tracer> TracerProvider::GetTracer(
+nostd::shared_ptr<opentelemetry::trace::Tracer> TracerProvider::GetTracer(
     nostd::string_view library_name,
     nostd::string_view library_version) noexcept
 {
-  return opentelemetry::nostd::shared_ptr<opentelemetry::trace::Tracer>(tracer_);
+  // if (library_name == "") {
+  //   // TODO: log invalid name.
+  // }
+
+  const std::lock_guard<std::mutex> guard(lock_);
+
+  auto lib = InstrumentationLibrary::create(library_name, library_version);
```

Comment:
If `GetTracer` were to be used in an instrumentation to retrieve a tracer when creating spans, this potentially creates 3 allocations and deallocs on a hot path.

Would it make sense to avoid heap allocations here for the usual case when a tracer already exists?

I mean using `GetTracer()` every time could perhaps be considered bad use of the API as opposed to caching a tracer in the instrumentation, but I think it might be too easy too misuse the API and just default to `GetTracer()`

Reply:
@seemk thanks. I think this is a valid concern. I updated the PR and delayed the heap allocation to the time when creating new `Tracer` is necessary.